sycl/doc/KernelProgramCache.md

```diff
@@ -236,13 +301,100 @@ thread will try to build the same object once more.
 
 <a name="exception-data">2</a>: Actually, we store contents of the exception:
 its message and error code.
 
+### Persistent cache storage structure
+
+The device code image are stored on file system using structure below:
+```
+<cache_root>/
+	<device_hash>/
+		<device_image_hash>/
+			<spec_constants_values_hash>/
+				<build_options_hash>/
+					<n>.src
+					<n>.bin
+```
+ - `<cache_root>` - root directory storing cache files;
+ - `<device_hash>` - hash out of device information used to identify target device;
+ - `<device_image_hash>` - hash made out of device image used as input for the JIT compilation;
+ - `<spec_constants_values_hash>` - hash for specialization constants values;
+ - `<build_options_hash>` - hash for all build options;
+ - `<n>` - sequential number of hash collisions. When hashes matches for the specific build but full values don't, new cache item is added with incremented value (enumeration started from 0).
+
+Two files per cache item are stored on disk:
+ - `<n>.src` contains full values for build parameters (device information, specialization constant values, build options, device image) which is used to resolve hash collisions and analysis of cached items. 
+ - `<n>.bin` contains built device code.
+
+### Inter-process safety
+
+For on-disc cache there might be access collisions for accessing the same file
+from different instance of SYCL applications:
+- write collision happens when 2 instances of the same application are started
+to write to the same cache file/directory;
+- read collision may happen if one application is writing to the file and the
+other instance of the application is trying to read from it while write
+operation is not finished.
+
+To avoid collisions the file system entries are locked for read-write access
+until write operation is finished. e.g if new file or directory should be
+created/deleted parent directory is locked, file is created in locked state,
+then the directory and the file are unlocked.
+
+To address cases with high lock rate (multiple copies of the SYCL applications
+are run in parallel and use the same cache directory) nested directories
+representing cache keys are used to minimize locks down to applications which
+build the same device with the same parameters. Directory is locked for minimum
+time because it can be unlocked once subdirectory is created. If file is created in
+directory the directory should be locked until file creation done.
+
+Advisory locking <sup>[1](#advisory-lock)</sup> is used to ensure that the
+user/OS tools are able to manage files.
+
+<a name="advisory-lock">1.</a> Advisory locks work only when a process
+explicitly acquires and releases locks, and are ignored if a process is not aware
+of locks.
+
+### Cache eviction mechanism
+
+Cache eviction mechanism is required to avoid resources overfloat both for
+memory and disk. The general idea is to delete items following LRU (least
+recently used) strategy both for in-memory and persistent cache.
+
```

Comment:
LRU can be quite tricky to implement for the in-memory case, so I would like to see more detailed design in this doc. Can we add TODO for now?

Reply:
added TODO